From 2df626843fe2c548241a8795d306770538ee9c98 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Sat, 24 Jan 2026 00:15:44 +0100
Subject: [PATCH 03/12] HACK: kernel32: Lock GetProcAddress hack to when
 needed.

Can also be enabled with PROTON_ENABLE_INT3_HACK=1.
---
 dlls/kernel32/module.c | 33 +++++++++++++++++++++++++++++++--
 1 file changed, 31 insertions(+), 2 deletions(-)

diff --git a/dlls/kernel32/module.c b/dlls/kernel32/module.c
index 6c0152c9ca6..461f57ae627 100644
--- a/dlls/kernel32/module.c
+++ b/dlls/kernel32/module.c
@@ -262,7 +262,36 @@ BOOL WINAPI GetBinaryTypeA( LPCSTR lpApplicationName, LPDWORD lpBinaryType )
 }
 
 #ifdef __x86_64__
-/* workaround for tpshell */
+static BOOL needs_int3_hack(void)
+{
+    static volatile int cache = -1;
+    
+    if (cache == -1)
+    {
+        const WCHAR *p, *name = NtCurrentTeb()->Peb->ProcessParameters->ImagePathName.Buffer;
+        WCHAR env[8];
+        BOOL ret;
+        
+        if ((p = wcsrchr(name, '/')))
+            name = p + 1;
+        if ((p = wcsrchr(name, '\\')))
+            name = p + 1;
+        
+        ret = ((!wcsicmp(name, L"Endfield.exe")) ||
+               (!wcsicmp(name, L"EM-Win64-Shipping.exe")));
+        
+        if (GetEnvironmentVariableW(L"PROTON_ENABLE_INT3_HACK", env, ARRAY_SIZE(env)))
+        {
+            if (_wtoi(env) == 1)
+                ret = TRUE;
+        }
+        
+        cache = ret;
+    }
+    
+    return cache;
+}
+
 static void __attribute__((naked)) int3_stub( void )
 {
     asm("int3\t\n"
@@ -296,7 +325,7 @@ FARPROC get_proc_address( HMODULE hModule, LPCSTR function )
         ANSI_STRING     str;
 
 #ifdef __x86_64__
-        if (strcmp( function, "KiUserApcDispatcher" ) == 0 || strcmp( function, "KiUserCallbackDispatcher" ) == 0)
+        if (needs_int3_hack() && (strcmp( function, "KiUserApcDispatcher" ) == 0 || strcmp( function, "KiUserCallbackDispatcher" ) == 0))
         {
             FIXME( "HACK: returning int3 stub instead of %s\n", function );
             return (FARPROC)&int3_stub;

-- 
2.54.0

